pkg/daemon/ceph/client/upgrade.go

```diff
@@ -104,6 +157,334 @@ func EnableNautilusOSD(context *clusterd.Context) error {
+			version, err := cephver.ExtractCephVersion(v)
+			if err != nil {
+				return vv, fmt.Errorf("failed to extract ceph version. %+v", err)
+			}
+			vv = *version
+			// break right after the first iteration
+			// the first one is always the least up-to-date
+			break
+		}
+	}
+
+	return vv, nil
+}
+
+func daemonMapEntry(versions *CephDaemonsVersions, daemonType string) (map[string]int, error) {
+	switch daemonType {
+	case "mon":
+		return versions.Mon, nil
+	case "mgr":
+		return versions.Mgr, nil
+	case "osd":
+		return versions.Osd, nil
+	case "rgw":
+		return versions.Rgw, nil
+	case "mirror":
+		return versions.RbdMirror, nil
+	}
+
+	return nil, fmt.Errorf("invalid daemonType %s", daemonType)
+}
+
+func allOSDsSameHost(context *clusterd.Context, clusterName string) (bool, error) {
+	tree, err := HostTree(context, clusterName)
+	if err != nil {
+		return false, fmt.Errorf("failed to get the osd tree %+v", err)
+	}
+
+	osds, err := OsdListNum(context, clusterName)
+	if err != nil {
+		return false, fmt.Errorf("failed to get the osd list %+v", err)
+	}
+
+	hostOsdTree := buildHostListFromTree(tree)
+	hostOsdNodes := len(hostOsdTree.Nodes)
+	if hostOsdNodes == 0 {
+		return false, fmt.Errorf("no host in crush map yet?")
+	}
+
+	// If the number of OSD node is 1, chances are this is simple setup with all OSDs on it
+	if hostOsdNodes == 1 {
+		// number of OSDs on that host
+		hostOsdNum := len(hostOsdTree.Nodes[0].Children)
+		// we take the total number of OSDs and remove the OSDs that are out of the CRUSH map
+		osdUp := len(osds) - len(tree.Stray)
+		// If the number of children of that host (basically OSDs) is equal to the total number of OSDs
+		// We can assume that all OSDs are running on the same machine
+		if hostOsdNum == osdUp {
+			return true, nil
+		}
+	}
+
+	return false, nil
+}
+
+func buildHostListFromTree(tree OsdTree) OsdTree {
+	var osdList OsdTree
+
+	for _, t := range tree.Nodes {
+		if t.Type == "host" {
+			osdList.Nodes = append(osdList.Nodes, t)
+		}
+	}
+
+	return osdList
+}
+
+// osdDoNothing determines wether we should perfom upgrade pre-check and post-checks for the OSD daemon
+// it checks for various cluster info like number of OSD and their placement
+// it returns 'true' if we need to do nothing and false and we should pre-check/post-check
+func osdDoNothing(context *clusterd.Context, clusterName string) bool {
+	versions, err := GetCephVersions(context, clusterName)
+	if err != nil {
+		logger.Warningf("failed to get ceph daemons versions. %+v, this likely means there is no cluster yet.", err)
+		return true
+	}
+
+	if len(versions.Osd) == 1 {
+		// now trying to parse and find how many osds are presents
+		// if we have less than 3 osds we skip the check and do best-effort
+		for _, osdCount := range versions.Osd {
+			if osdCount < 3 {
+				logger.Warningf("the cluster has less than 3 OSDs, not performing upgrade check, running in best-effort")
+				return true
+			}
+		}
+
+		// aio means all in one
```

Comment:
why not just spell it out to allInOne?

Reply:
`allInOne` is not pretty to read and `aio` shorter. That's what I thought.